runtime/src/error.c

```diff
@@ -45,19 +49,52 @@
 #include <libunwind.h>
 
 #ifdef __linux__
+#include <dlfcn.h> // for dladdr
 // We create a pipe with addr2line and try to get a line number
 // Currently the precise line number works only on linux64
-static int chpl_unwind_getLineNum(char *process_name, void *addr){
-  char buf[128];
+static int chpl_unwind_getLineNum(void *addr){
+
+  // Compute the object containing the address
+  int rc;
+  Dl_info info;
+  intptr_t relativeAddr;
+  char process_name[256];
+  char buf[256];
+  const char* path;
   int line;
   char* p;
   FILE *f;
 
+  rc = dladdr(addr, &info);
+  if (rc == 0)
+    return 0; // dladdr failed.
+
+  // Compute the relative address within the object
+  relativeAddr = (intptr_t)addr - (intptr_t)info.dli_fbase;
+
+  // Compute the path to the file containing the object
+  path = info.dli_fname;
+  if (path == NULL) {
+    // Try using the file path from the current executable
+    ssize_t len;
+    len = readlink("/proc/self/exe", process_name, sizeof(process_name));
```

Comment:
There are circumstances in which `readlink()` could return -1, so that case needs to be handled.